manifests/pulp.pp

```diff
@@ -129,50 +134,85 @@
-    db_ssl                 => $mongodb_ssl,
-    db_ssl_keyfile         => $mongodb_ssl_keyfile,
-    db_ssl_certfile        => $mongodb_ssl_certfile,
-    db_verify_ssl          => $mongodb_verify_ssl,
-    db_ca_path             => $mongodb_ca_path,
-    db_unsafe_autoretry    => $mongodb_unsafe_autoretry,
-    db_write_concern       => $mongodb_write_concern,
-    manage_db              => $manage_mongodb,
+    server_name               => $server_name,
+    messaging_url             => "ssl://${katello::params::qpid_hostname}:5671",
+    messaging_ca_cert         => $certs::qpid_client::qpid_client_ca_cert,
+    messaging_client_cert     => $certs::qpid_client::qpid_client_cert,
+    messaging_transport       => 'qpid',
+    messaging_auth_enabled    => false,
+    broker_url                => "qpid://${katello::params::qpid_hostname}:5671",
+    broker_use_ssl            => true,
+    yum_max_speed             => $yum_max_speed,
+    manage_broker             => false,
+    manage_httpd              => $manage_vhost_standalone,
+    https_cert                => $https_cert,
+    https_key                 => $https_key,
+    https_ca_cert             => $https_ca_cert,
+    manage_plugins_httpd      => true,
+    manage_squid              => true,
+    enable_rpm                => $katello::params::enable_yum,
+    enable_iso                => $katello::params::enable_file,
+    enable_deb                => $katello::params::enable_deb,
+    enable_puppet             => $katello::params::enable_puppet,
+    enable_docker             => $katello::params::enable_docker,
+    enable_ostree             => $katello::params::enable_ostree,
+    num_workers               => $num_workers,
+    enable_parent_node        => false,
+    repo_auth                 => true,
+    puppet_wsgi_processes     => 1,
+    enable_katello            => true,
+    subscribe                 => Class['certs'],
+    worker_timeout            => $worker_timeout,
+    db_name                   => $mongodb_name,
+    db_seeds                  => $mongodb_seeds_real,
+    db_username               => $mongodb_username,
+    db_password               => $mongodb_password,
+    db_replica_set            => $mongodb_replica_set,
+    db_ssl                    => $mongodb_ssl or $mongodb_server_ssl,
+    db_ssl_keyfile            => $mongodb_ssl_keyfile_real,
+    db_ssl_certfile           => $mongodb_ssl_certfile_real,
+    db_verify_ssl             => $mongodb_verify_ssl,
+    db_ca_path                => $mongodb_ca_path_real,
+    db_unsafe_autoretry       => $mongodb_unsafe_autoretry,
+    db_write_concern          => $mongodb_write_concern,
+    manage_db                 => $manage_mongodb,
+    mongodb_bind_ip           => $mongodb_bind_ip,
+    mongodb_server_ssl        => $mongodb_server_ssl,
+    mongodb_server_ssl_bundle => $mongodb_server_ssl_bundle,
+    mongodb_server_ssl_ca     => $mongodb_server_ssl_ca,
+    mongodb_ipv6              => $expose_mongodb_ipv6,
```

Comment:
Rather than passing the parameters, I'd prefer to explicitly define `mongodb::server` just before the `pulp` class.

Reply:
I'm going to need some help understanding *why* that design choice is better. And, probably then, why we have an `include mongodb::server` in the pulp module at all. I have a hard time following the design granularities and when to use what.

Reply:
The reasoning is that you can only manage a class once. Both the MongoDB and PostgreSQL services are classes. In regular programming languages you'd refer to them as singletons. It's not possible to instantiate multiple instances. If multiple services depend on a service (think both Foreman and Pulpcore that need PostgreSQL) then it's a conflict. That's why Foreman only includes PostgreSQL and ensures the correct chaining. (first DB, then DB migrations). The same thing with Pulpcore. `include` is what a `.getInstance()` would be in other languages where `class myclass {}` is like calling the constructor.

To managing it, there are 2 solutions: hiera (what we in the installer do for PostgreSQL) or calling the constructor before ever including it. Here you'd take the latter approach.

Typically this is reserved for profile modules which combine multiple basic modules into a specific deployment. I could puppet-katello as a profile module.

Reply:
Can these two lines (https://github.com/theforeman/puppet-pulp/blob/master/manifests/database.pp#L7-L8) be defined with the `include mongodb::server` ?

Reply:
No, `include` can't have a require statement. `include` really means "ensure it's somewhere in the catalog" but makes no guarantees whatsoever about ordering. In that particular case, I think `require mongodb::server` would be the statement you're looking for.